CollectInsertSizeMetricsSpark addressed comments from previous review

SHuang-Broad opened the pull request:
Hi Adam,

I've reformatted the code and fixed some issues you raised in the first review.
Specifically:
1. Default values for tool's boolean arguments are changed to false.
2. Created an (ugly) SV program group.
3. Uses Apache library to calculate statistics.

Somethings are still to be changed:
1. A lot of todo's.
2. Histogram plot is created by Picard's R script, which was copied to SV.

Of particular concern is that Picard's collection of the insert size distribution is potentially biased and has edge case bugs, see issue #1519. So I've used the following R script (can't attach on github) for collecting stats on the test bam file.

    # Serves to collect basic metrics on insert sizes of test bam file and use for 
    # unit testing assertions.
    testBamFile <- "insert_size_metrics_test.bam"

    library(Rsamtools)
    # take only first end of read pairs
    filter <- ScanBamParam(what=scanBamWhat(),flag=scanBamFlag(isFirstMateRead=T))
    dd <- scanBam(testBamFile, param = filter)
    dd <- dd[[1]]
    sizes <- abs(dd$isize[!is.na(dd$isize)])
    sizes <- sort(sizes)

    metrics <- list()
    metrics[["count"]] <- length(sizes)
    metrics[["median"]] <- median(sizes)
    metrics[["mean"]] <- mean(sizes)
    metrics[["min"]] <- min(sizes)
    metrics[["max"]] <- max(sizes)
    metrics[["MAD"]] <- mad(sizes, constant = 1)
    metrics[["SD"]] <- sd(sizes)

    q <- quantile(sizes, probs = c(0.005, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45,
                                   0.55, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 0.995))
    metrics[["99_percent_width"]] <- as.integer(round(q[20] - q[1]))
    metrics[["90_percent_width"]] <- as.integer(round(q[19] - q[2]))
    metrics[["80_percent_width"]] <- as.integer(round(q[18] - q[3]))
    metrics[["70_percent_width"]] <- as.integer(round(q[17] - q[4]))
    metrics[["60_percent_width"]] <- as.integer(round(q[16] - q[5]))
    metrics[["50_percent_width"]] <- as.integer(round(q[15] - q[6]))
    metrics[["40_percent_width"]] <- as.integer(round(q[14] - q[7]))
    metrics[["30_percent_width"]] <- as.integer(round(q[13] - q[8]))
    metrics[["20_percent_width"]] <- as.integer(round(q[12] - q[9]))
    metrics[["10_percent_width"]] <- as.integer(round(q[11] - q[10]))

akiezun commented on src/main/java/org/broadinstitute/hellbender/cmdline/programgroups/StructuralVariationProgramGroup.java:
```diff
@@ -0,0 +1,16 @@
+
+import org.broadinstitute.hellbender.cmdline.CommandLineProgramGroup;
+
+public final class StructuralVariationProgramGroup implements CommandLineProgramGroup {
```
This class needs a javadoc

akiezun commented on src/main/java/org/broadinstitute/hellbender/cmdline/programgroups/StructuralVariationProgramGroup.java:
```diff
@@ -0,0 +1,16 @@
+
+    @Override
+    public String getName() {
+        return "Spark tools for structural variation analysis";
```
either add Spark to the class name or remove Spark from this string and the description string.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSpark.java:
according to java style guide https://google.github.io/styleguide/javaguide.html#s4.8.1-enum-classes enum classes should be named `LikeThis` and the constants (enum objects) should be named `LIKE_THIS`

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSpark.java:
fix the doc - 0,1,2 are not valid values anymore because the enum exists

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSpark.java:
well, this is impossible here `filteredReads == null` because the line above would have blown up.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSpark.java:
also the contract to `getReads` is that it never returns null so you dont need to check it here

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSpark.java:
what is this `b` for?

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSpark.java:
no reason to have this `return` here

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
delete dead code

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
why is this RDD as a field? The code does not seem to use it

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
this code here should be extracted to its own method. This constructor is way too long

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
remember that if you use an RDD twice it will be recomputed again unless you cache it!

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
replace this whole method with a 1 liner:
```
 rdd.collect().stream().mapToDouble(i -> (double) i).toArray()
```

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
use interfaces in declarations, so:
```
final SortedMap<Integer, Long> hist = new TreeMap<>(temp);
```

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
rename the field to `metrics`

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVCustomReadFilter.java:
either pass these 6 arguments separately or make a new class for them. Passing a 6-element array is not good.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVCustomReadFilter.java:
create the `pfilter.and(sfilter)` object in the constructor and only call `test` on that object here - otherwise your creating the same object over and over, once for each read

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVCustomReadFilter.java:
same issue - expand the options or make a new class

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVCustomReadFilter.java:
what is the `1` or `2` for? These int values are not meaningful.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVCustomReadFilter.java:
the `sixOptions` must be expanded into 6 different fields with meaningful names or a class with 6 fields with meaningful names

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVCustomReadFilter.java:
`private static final class`

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSparkUnitTest.java:
why are 2 tests commented out?

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSparkUnitTest.java:
was this supposed to be cram?

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSparkUnitTest.java:
do tests really fail? If yes, then we must fix them

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
make two separate locals with meaningful names

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
don't reuse locals without a very good reason (there's no reason here)

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
here too, don't reuse locals without a very good reason (there's no reason here)

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
why is this called `htslib`Histogram?

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
inline the temp - it's only used in 1 place

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVCustomReadFilter.java:
If this class is going to be used in multiple SV tools, then you need to add unit tests for it. Otherwise, make it a private class inside `InsertSizeMetricsCollectorSpark`

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/SVCustomReadFilter.java:
there's no reason to make this a static field

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSparkUnitTest.java:
Use this method instead everywhere to assert equality of doubles
```
Assert.assertEquals(metrics.MEAN_INSERT_SIZE, 40.1, DOUBLE_TOLERANCE);
```

akiezun commented:
@SHuang-Broad looking better - I added a bunch of new comments. Please assign back to me when you've addressed them

SHuang-Broad commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSparkUnitTest.java:
They are not used:
1. sam is the same as bam (remember the empty bam is regenerated from sam?) so testing it is equivalent to testing other parts of GATK
2. the case with reference is not yet implemented or checked--but planned to be.

SHuang-Broad commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSparkUnitTest.java:
Aren't they the same? The cram unit test file is empty, so essentially it is just another sam file, like the bam file case.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
It was used on lines 81, 91, and 98. But if you think recomputing the RDD is a better tradeoff, I'll remove it and recompute when necessary.

SHuang-Broad commented:
Hi Adam,

I've addressed your comments in the updated branch.
The biggest change is that SVCustomReadFilter is now merged into CollectInsertSizeMetricsSpark, as reads will mostly likely be filtered by k-mers later, per my understanding of Chris and Ted's plan.

Thank you.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
yes, i think storing this as a field does not make sense. The reason is that after the InsertSizeMetricsCollectorSpark object is created you never change the state of this field and you never need to even read this field, right? In other words, this field should really be a temporary variable inside of the constructor.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
should be a static method - or use the 1 liner I wrote here before

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
don't reuse the `percent` temporary variable

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
dont reuse temporary variables

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
replace by:
`final SortedMap<Integer, Long> hist = new TreeMap<>(lengthRDD.countByValue());`

and eliminate `temp`

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
split into two declarations

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
what is `10` for? Is is the same `10` as on the previous line?

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
split declarations

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
split this - dont write multiple lines on one line

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
declare as `long` (lowercase)

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
The part of this method that takes a `SortedMap<Integer, Long> hist ` and returns a `long[]` should be extracted to a separate method and unit-tested separately

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
can sz1 or sz2 ever be null given that they come from a `countByValue()` ?

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
split this array into 2 temportary variables

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
final `int sz: hist.keySet()` <- better yet, rename `sz` to something better like `size`

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java:
use SortedMap in the declaration and remove the `temp`

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSparkUnitTest.java:
well, then fix the cram file and use it here (if cram on spark works now)

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/CollectInsertSizeMetricsSparkUnitTest.java:
in that case, delete the unused code.

akiezun commented:
getting there - a  few more issues with the histogram computations. Back to @SHuang-Broad

SHuang-Broad commented:
Addressed comments, 3rd round.
Thank you!
I apologize for taking so many rounds.

akiezun commented:
great work. only 1 more thing (unit tests for that one method) and it's ready.

SHuang-Broad commented:
Added test. 
Thanks!

akiezun commented:
the adam.log file should be removed

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GatherSplitReadsSimpleSpark.java:
i think this class got here by mistake right?

akiezun commented:
looks good! 2 bogus files got added by mistake. Can you remove them and then rebase and squash?

SHuang-Broad commented:
Done removing bogus files and rebase & squash. Please review.
Thanks!

akiezun commented:
:+1: looks great. thanks! sorry it took so many rounds - mostly stylistic issues, will be much easier next time

SHuang-Broad commented:
No problem. I'd rather take styling issues ironed out from the beginning, and in fact I learn quite a lot about GATK/Picard and Java in the process.
Thank you very much for your patience!

akiezun commented:
you're welcome and welcome to Hellbender/GATK4! congrats on your first contribution
